src/Adapters/Auth/oauth2.js

```diff
@@ -0,0 +1,231 @@
+/*
+ * This auth adapter is based on the OAuth 2.0 Token Introspection specification.
+ * See RFC 7662 for details (https://tools.ietf.org/html/rfc7662).
+ * It's purpose is to validate OAuth2 access tokens using the OAuth2 provider's
+ * token introspection endpoint (if implemented by the provider).
+ *
+ * The adapter accepts the following config parameters:
+ *
+ * 1. "tokenIntrospectionEndpointUrl" (string, required)
+ *      The URL of the token introspection endpoint of the OAuth2 provider that
+ *      issued the access token to the client that is to be validated.
+ *
+ * 2. "useridField" (string, optional)
+ *      The name of the field in the token introspection response that contains
+ *      the userid. If specified, it will be used to verify the value of the "id"
+ *      field in the "authData" JSON that is coming from the client.
+ *      This can be the "aud" (i.e. audience), the "sub" (i.e. subject) or the
+ *      "username" field in the introspection response, but since only the
+ *      "active" field is required and all other reponse fields are optional
+ *      in the RFC, it has to be optional in this adapter as well.
+ *      Default: - (undefined)
+ *
+ * 3. "appidField" (string, optional)
+ *      The name of the field in the token introspection response that contains
+ *      the appId of the client. If specified, it will be used to verify it's
+ *      value against the set of appIds in the adapter config. The concept of
+ *      appIds comes from the two major social login providers
+ *      (Google and Facebook). They have not yet implemented the token
+ *      introspection endpoint, but the concept can be valid for any OAuth2
+ *      provider.
+ *      Default: - (undefined)
+ *
+ * 4. "appIds" (array of strings, optional)
+ *      A set of appIds that are used to restrict accepted access tokens based
+ *      on a specific field's value in the token introspection response.
+ *      Default: - (undefined)
+ *
+ * 5. "authorizationHeader" (string, optional)
+ *      The value of the "Authorization" HTTP header in requests sent to the
+ *      introspection endpoint.
+ *      Eg. "Basic dXNlcm5hbWU6cGFzc3dvcmQ="
+ *
+ * 6. "debug" (boolean, optional)
+ *      Enables extensive logging using the "verbose" level.
+ *
+ * The adapter expects requests with the following authData JSON:
+ *
+ * {
+ *   "oauth2": {
+ *     "id": "user's OAuth2 provider-specific id as a string",
+ *     "access_token": "an authorized OAuth2 access token for the user",
+ *   }
+ * }
+ */
+
+import logger from '../../logger';
+const Https = require('https');
+const Parse = require('parse/node').Parse;
+const Path = require('path');
+const Url = require('url');
+const Querystring = require('querystring');
+
+const scriptName = Path.basename(__filename, '.js');
+
+// Returns a promise that fulfills if this user id is valid.
+function validateAuthData(authData, options) {
+  const _logger = _getLogger(validateAuthData.name);
```

Comment:
Also the full logger thing seems overkill, if it just prepends a string. (and it's untested as well).

You're only using verbose and error so all other methods should be removed.

Another possible improvement, when you get the logger, you probably want to pass the options, this way you don't need to test if logging is enabled around each calls but only in the logger calls.

Overall it seems overkill, to have this additional special logger that is never used anywhere else.

Also the logger is untested, so there's no way to know if it properly logs the objects and not `[Object object]`

Reply:
It seems that the logger was there to help you write the code and doesn't server a final actionable purpose in the future.

Reply:
You're right. It was there to help with development. I've removed it.

> Another possible improvement, when you get the logger

Thanks. It would be a good improvement indeed. But it was a better choice to remove the extra logging code.